workload/schemachange: improve errors; initial changes

jayshrivastava opened the pull request:
This PR contains the groundwork for error handling and what we want to do with random complex schema generation. I think this is a good point to stop and access if we want to continue on this path.

Simpler commits:
- workload/schemachange: improve abstraction for probabilistic error production 
This adds a global knob for which we can use to control errors due to existing/non-existing resources. This replaces a bunch of int literals. 

- workload/schemachange: allow schema to be specified when fetching a random Table
When calling `randTable` it is sometimes useful to provide a schema name. This commit adds this functionality. 
- workload/schemachange: fix malformed column and index names
Fixes a small bug were we would have names such as "colpublic.table1_1" instead of "col1_1". The latter is correct. 
- workload/schemachange: sync column names with tables
Fixes a small bug. We want columns for "tableN" to start with "colN", but there was a bug where we would use N+1. 

Refactoring:
- workload/schemachange: refactor worker into op builder 
Moves op generation logic from the schemaChangeWorker to a new struct

Error Handling:

- workload/schemachange: add expected error handling
adds states to the op generator for storing errors - it also updates the error handling code in the worker
 
- workload/schemachange: screen errors in addColumn op

- workload/schemachange: update views and tables to allow multiple source tables and views
(Also screen for errors in these ops)
 
- workload/schemachange: screen errors in addColumn op

jayshrivastava committed:
workload/schemachange: improve abstraction for probabilistic error production

Before, we would use int literals in several places the code to specify
how often we want a name conflict error to occur probabilisically. This change
refactors this behavior to use a global knob for organization purposes.

Release note: None

jayshrivastava committed:
workload/schemachange: allow schema to be specified when fetching a random table

Before, you could not specify the schema when calling randTable(). This gave us less
control over when errors occur. For example, when renaming a table, an error is
produced if you specify a different schema than the original table's schema. Now,
a schema can be specified to control when an error occurs in this case.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55521)
<!-- Reviewable:end -->

jayshrivastava committed:
workload/schemachange: fix malformed column and index names

Previously, the table name would sometimes get interpolated into
column or index names. For example, a name such as
"colpublic.table1_1" would be generated instead of "col1_1". This
commit fixes this behavior.

Release note: None

jayshrivastava committed:
workload/schemachange: sync column names with tables

It is desired that the columns of a table have the table
all have the table index number in the name. For example,
"colN_M" should exist on "tableN". Previously, there was a
double increment causing the numbers to misalign. This
commit fixes this behavior.

Release note: None

jayshrivastava committed:
workload/schemachange: refactor worker into op builder

Previously all operation generation logic was stateless and tied to the
schemaChangeWorker. Creating an independent builder struct will make it
easier to keep track of state when generating ops that relate to each other.

Release note: None

jayshrivastava commented on pkg/workload/schemachange/operation_generator.go:
```diff
@@ -0,0 +1,1132 @@
+		if !typeExists {
+			errCode = &pgcode.UndefinedObject
+		}
+	}
```
When the `NotNull` constraint is present, the behaviour with returning either a `NotNullViolation` or `DuplicateColumn` is pretty unpredictable. I thought I had it correct here, but eventually this happened:
```
root@localhost:26257/schemachange> create table foo(t TIMESTAMPTZ);
CREATE TABLE
Time: 80ms total (execution 80ms / network 0ms)

root@localhost:26257/schemachange> insert into foo values ('2016-06-22 19:10:25-07');
INSERT 1
Time: 43ms total (execution 43ms / network 0ms)

root@localhost:26257/schemachange> ALTER TABLE foo ADD COLUMN t TIMESTAMPTZ NOT NULL;
ERROR: null value in column "t" violates not-null constraint
SQLSTATE: 23502
```

In a way this is good because it adds test coverage. The downsides would be having to figure out exactly what causes each error and having to put this level of detail into each op. In this case, it may even be nondeterministic. It might be a good idea to return set of errors instead of one and checking that the resultant error exists in the set.

jayshrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55521)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/workload/schemachange/operation_generator.go, line 289 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-MJml5nPBfXwWYbrJ67Q:-MJml5nQ9fajbm17hRUk:bxngxba) ([raw file](https://github.com/cockroachdb/cockroach/blob/9403c199f3693beb54f5506182c15e3af0f6ab96/pkg/workload/schemachange/operation_generator.go#L289)):*
> ```Go
> 			errCode = &pgcode.UndefinedObject
> 		}
> 	}
> ```

When the `NotNull` constraint is present, the behaviour with returning either a `NotNullViolation` or `DuplicateColumn` is pretty unpredictable. I thought I had it correct here, but eventually this happened:
```
root@localhost:26257/schemachange> create table foo(t TIMESTAMPTZ);
CREATE TABLE
Time: 80ms total (execution 80ms / network 0ms)

root@localhost:26257/schemachange> insert into foo values ('2016-06-22 19:10:25-07');
INSERT 1
Time: 43ms total (execution 43ms / network 0ms)

root@localhost:26257/schemachange> ALTER TABLE foo ADD COLUMN t TIMESTAMPTZ NOT NULL;
ERROR: null value in column "t" violates not-null constraint
SQLSTATE: 23502
```

In a way this is good because it adds test coverage. The downsides would be having to figure out exactly what causes each error and having to put this level of detail into each op. In this case, it may even be nondeterministic. It might be a good idea to return set of errors instead of one and checking that the resultant error exists in the set.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
I think I'm cool with this refactor though I'd love to get more insight into where you plan on taking it. Right now, it doesn't feel like you're getting a lot out of lifting the functionality onto this generator struct. What do you have in mind for its future?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-:-MK5GIFK3inWZYjVhSPb:b5nus67)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jayshrivastava)

___
*[pkg/workload/schemachange/operation_generator.go, line 127 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-MK5Fe2a5FJQ2EJ0m39i:-MK5Fe2a5FJQ2EJ0m39j:bwjuskd) ([raw file](https://github.com/cockroachdb/cockroach/blob/9403c199f3693beb54f5506182c15e3af0f6ab96/pkg/workload/schemachange/operation_generator.go#L127)):*
> ```Go
> // stochastic attempts and if verbosity is >= 2 the unsuccessful attempts are
> // recorded in `log` to help with debugging of the workload.
> func (og operationGenerator) randOp(tx *pgx.Tx) (string, string, *pgcode.Code, error) {
> ```

my sense is that this should be a method on the pointer, otherwise you can't modify `og`. I get that right now you aren't trying to but moving forward I suspect you will be trying to.


<!-- Sent from Reviewable.io -->

jayshrivastava committed:
workload/schemachange: add expected error handling

Previously, there was no mechanism for checking for
expected errors vs unexpected errors when operations fail
due to intentionally produced errors. This commit adds logic for
screening the types of errors that are generated and state variables
to track which errors are expected.

Release note: None

jayshrivastava committed:
workload/schemachange: screen errors in addColumn op

This change updates the addColumn op to screen for potential errors.
The op now updates the state of the op generator so that expected errors
can safely be ignored.

Release note: None

jayshrivastava committed:
workload/schemachange: update views and tables to allow multiple source tables or views

Previously, the ops createView and createTableAs would only create resources
based on one table. This commit updates those ops so that they allow for multiple
souce views and/or tables. It also adds error screening to the ops.

Release note: None

jayshrivastava committed:
workload/schemachange: screen for errors in createTable op

Updates error screening so that expected errors are explicit
and safely ignored in createTable op

Release note: None

ajwerner reviewed (changes requested):
This is very close. Let's get it merged soon. Mostly just nits about the errorCodeSet. 

> Reviewed 1 of 3 files at r13.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-:-MKe6GS35I37XVmpQ_zw:b-8nbdol)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jayshrivastava)

___
*[pkg/workload/schemachange/error_code_set.go, line 31 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-MKe36kQ5DheJX_UqXxU:-MKe36kQ5DheJX_UqXxV:b-qtnqrj) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c9c5fa03488bd91da5fd5aff305f488030c40bc/pkg/workload/schemachange/error_code_set.go#L31)):*
> ```Go
> }
> 
> func (set *errorCodeSet) addIdempotent(code pgcode.Code) {
> ```

nit: I don't think `Idempotent` adds a lot to the name of this method.
___
*[pkg/workload/schemachange/error_code_set.go, line 31 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-MKe3qIk5wVp3W5Gjy87:-MKe3qIk5wVp3W5Gjy88:bac0pg6) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c9c5fa03488bd91da5fd5aff305f488030c40bc/pkg/workload/schemachange/error_code_set.go#L31)):*
> ```Go
> }
> 
> func (set *errorCodeSet) addIdempotent(code pgcode.Code) {
> ```

These don't need to be pointers, maps are already reference data types (i.e. a map value is mutable).
___
*[pkg/workload/schemachange/error_code_set.go, line 35 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-MKe463d3SgZD37HaRdm:-MKe463d3SgZD37HaRdn:bgs55g9) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c9c5fa03488bd91da5fd5aff305f488030c40bc/pkg/workload/schemachange/error_code_set.go#L35)):*
> ```Go
> }
> 
> func (set *errorCodeSet) addIdempotentWithConditions(codesWithConditions []codeWithCond) {
> ```

nit: lose the idempotent, I think set implies it. There's also something 
___
*[pkg/workload/schemachange/error_code_set.go, line 44 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-MKe3kz21p0UsfxPKER9:-MKe3kz21p0UsfxPKERA:bx8sp6l) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c9c5fa03488bd91da5fd5aff305f488030c40bc/pkg/workload/schemachange/error_code_set.go#L44)):*
> ```Go
> 
> func (set *errorCodeSet) reset() {
> 	*set = map[pgcode.Code]bool{}
> ```

nit: it's better to do the delete, the go compiler detects the idiom and does an efficient clear: https://github.com/golang/go/issues/20138

```
for k := range *set {
    delete(*set, k)
}
```
___
*[pkg/workload/schemachange/error_screening.go, line 33 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-MKe5Sm13EKjDcz-93Pp:-MKe5Sm13EKjDcz-93Pq:b-67duvo) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c9c5fa03488bd91da5fd5aff305f488030c40bc/pkg/workload/schemachange/error_screening.go#L33)):*
<details><summary><i alt="summary">Quoted 11 lines of code…</i></summary>

> ```
> 	q := fmt.Sprintf(`SELECT EXISTS (
> 	SELECT table_name
>     FROM information_schema.tables
>    WHERE table_schema = '%s'
>      AND table_name = '%s'
>    )`, tableName.SchemaName, tableName.ObjectName.String())
> 	var exists bool
> 	if err := tx.QueryRow(q).Scan(&exists); err != nil {
> 		return false, err
> 	}
> 	return exists, nil
> ```
</details>

feels like you could extract a helper that takes a query sql (and maybe its args, you know you don't need to always string format the query, right?), and returns a bool and error. Might help some with the repetition her. 


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55521)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jayshrivastava)

___
*[pkg/workload/schemachange/error_code_set.go, line 35 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-MKe463d3SgZD37HaRdm:-MKfQkQHBygGplSz2NNm:b-49flm0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c9c5fa03488bd91da5fd5aff305f488030c40bc/pkg/workload/schemachange/error_code_set.go#L35)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: lose the idempotent, I think set implies it. There's also something 
</blockquote></details>

I don't think you finished the comment. I changed the name. I also think this fcn is useful for saving some space compared to having several if statements in a row with `set.add()`
```
if .... {
    set.add(code)
}
```
___
*[pkg/workload/schemachange/error_screening.go, line 33 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-MKe5Sm13EKjDcz-93Pp:-MKfRPZG7tmgRXWZDK-V:bnrml81) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c9c5fa03488bd91da5fd5aff305f488030c40bc/pkg/workload/schemachange/error_screening.go#L33)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	q := fmt.Sprintf(`SELECT EXISTS (
> 	SELECT table_name
>     FROM information_schema.tables
>    WHERE table_schema = '%s'
>      AND table_name = '%s'
>    )`, tableName.SchemaName, tableName.ObjectName.String())
> 	var exists bool
> 	if err := tx.QueryRow(q).Scan(&exists); err != nil {
> 		return false, err
> 	}
> 	return exists, nil
> ```
feels like you could extract a helper that takes a query sql (and maybe its args, you know you don't need to always string format the query, right?), and returns a bool and error. Might help some with the repetition her. 
</blockquote></details>

I'm not exactly sure what you mean here. I felt that storing the SQL string templates inside these functions and interpolating the args with `fmt.Sprintf` reduces duplication because we only store the string templates in one place. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55521)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jayshrivastava)

___
*[pkg/workload/schemachange/error_code_set.go, line 35 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-MKe463d3SgZD37HaRdm:-MKf_Ee2E3vN4XaU24X4:b-ee3gwp) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c9c5fa03488bd91da5fd5aff305f488030c40bc/pkg/workload/schemachange/error_code_set.go#L35)):*
<details><summary><i>Previously, jayshrivastava (Jayant Shrivastava) wrote…</i></summary><blockquote>

I don't think you finished the comment. I changed the name. I also think this fcn is useful for saving some space compared to having several if statements in a row with `set.add()`
```
if .... {
    set.add(code)
}
```
</blockquote></details>

I agree with that. It's a handy construct. I was wondering whether I liked it as a method and I decided I couldn't really fault it. I was thinking maybe it's be better to invert it so that it's the slice with the method:

```
type codesWithConditions []struct {
    code         pgcode.Code
    condition bool
}

func (c codesWithConditions) add(s errorCodeSet) {
     for _, cc := range c {
         if cc.condition {
              s.add(cc.code)
         }
     }
}

```
```
   codesWithConditions{
       {code: ..., condition: ...},
   }.add(s)

```
___
*[pkg/workload/schemachange/error_code_set.go, line 52 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-MKfYNOO9Ccv_QhpMQ2j:-MKfYNOO9Ccv_QhpMQ2k:b8rhihi) ([raw file](https://github.com/cockroachdb/cockroach/blob/54da2eac64848bebbe738a52163e2421f096ae2c/pkg/workload/schemachange/error_code_set.go#L52)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> 	if _, ok := set[code]; ok {
> 		return true
> 	}
> 	return false
> ```
</details>

total nit:

```
_, ok := set[code]
return ok
```
___
*[pkg/workload/schemachange/error_code_set.go, line 61 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-MKfYSxZ5oOHQdDMqSIs:-MKfYSxZ5oOHQdDMqSIt:b-aw9jns) ([raw file](https://github.com/cockroachdb/cockroach/blob/54da2eac64848bebbe738a52163e2421f096ae2c/pkg/workload/schemachange/error_code_set.go#L61)):*
<details><summary><i alt="summary">Quoted 6 lines of code…</i></summary>

> ```
> 	var codes []string
> 	for code := range set {
> 		codes = append(codes, code.String())
> 	}
> 	return strings.Join(codes, ",")
> }
> ```
</details>

nits: probably you'll want this to be deterministic, so doing a `sort.Strings` beforehand might be nice
___
*[pkg/workload/schemachange/error_code_set.go, line 65 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-MKfYcTs3UagHhDFFPNi:-MKfYcTs3UagHhDFFPNj:bm9qp7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/54da2eac64848bebbe738a52163e2421f096ae2c/pkg/workload/schemachange/error_code_set.go#L65)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> 	for range set {
> 		return false
> 	}
> 	return true
> ```
</details>

nit: `return len(set) == 0`
___
*[pkg/workload/schemachange/error_screening.go, line 33 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-MKe5Sm13EKjDcz-93Pp:-MKfYgKm3Gao0o66eHaX:bavgcnp) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c9c5fa03488bd91da5fd5aff305f488030c40bc/pkg/workload/schemachange/error_screening.go#L33)):*
I meant something like: 
```
func scanBool(tx *pgx.Tx, query string, args ...interface) (b bool, err error) {
    err = tx.QueryRow(query, args...).Scan(&b)
    return b, err
}
```

then `tableExists` could be:

```
func tableExists(tx *pgx.Tx, tableName *tree.TableName) (bool, err) {
    return scanBool(`
SELECT EXISTS(
        SELECT table_name
          FROM information_schema.tables
         WHERE table_schema = $1 AND table_name = $2
       );
`, tableName.SchemaName, tableName.ObjectName.String())

}
```


---

> I felt that storing the SQL string templates inside these functions and interpolating the args with fmt.Sprintf reduces duplication because we only store the string templates in one place.

I'm not sure I understand the value of the `Sprintf` vs placeholders.


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55521)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/workload/schemachange/error_code_set.go, line 31 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-MKe3qIk5wVp3W5Gjy87:-MKjGt9h97HzBReO6lig:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c9c5fa03488bd91da5fd5aff305f488030c40bc/pkg/workload/schemachange/error_code_set.go#L31)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

These don't need to be pointers, maps are already reference data types (i.e. a map value is mutable).
</blockquote></details>

Done.
___
*[pkg/workload/schemachange/error_code_set.go, line 52 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-MKfYNOO9Ccv_QhpMQ2j:-MKjGuiHEFNeXHwhy2PV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/54da2eac64848bebbe738a52163e2421f096ae2c/pkg/workload/schemachange/error_code_set.go#L52)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	if _, ok := set[code]; ok {
> 		return true
> 	}
> 	return false
> ```
total nit:

```
_, ok := set[code]
return ok
```
</blockquote></details>

Done.
___
*[pkg/workload/schemachange/error_code_set.go, line 61 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-MKfYSxZ5oOHQdDMqSIs:-MKjGw4Z0GXKwvdRTSIV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/54da2eac64848bebbe738a52163e2421f096ae2c/pkg/workload/schemachange/error_code_set.go#L61)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	var codes []string
> 	for code := range set {
> 		codes = append(codes, code.String())
> 	}
> 	return strings.Join(codes, ",")
> }
> ```
nits: probably you'll want this to be deterministic, so doing a `sort.Strings` beforehand might be nice
</blockquote></details>

Done.
___
*[pkg/workload/schemachange/operation_generator.go, line 127 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-MK5Fe2a5FJQ2EJ0m39i:-MKjH-O386yg36qxG9PJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9403c199f3693beb54f5506182c15e3af0f6ab96/pkg/workload/schemachange/operation_generator.go#L127)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

my sense is that this should be a method on the pointer, otherwise you can't modify `og`. I get that right now you aren't trying to but moving forward I suspect you will be trying to.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Thanks for iterating on this with me. I hope you're pleased with where it ended up. 

> Reviewed 1 of 3 files at r22, 1 of 4 files at r28, 1 of 3 files at r29, 2 of 3 files at r30, 1 of 1 files at r31.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-:-MKjJ3mk-nrY_wWkyXXk:b5l990w)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
Ofc! Thank you for reviewing it thoroughly. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55521#-:-MKjJLpN4--J_jqEc2zA:b-wpc1i0)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

jayshrivastava commented:
bors r=ajwerner
